content/blog/2020/01/2020-01-23-ui-plugins.adoc

```diff
@@ -0,0 +1,708 @@
+In order to use such a table in your plugin view you need to create the table in the associated
+Jelly file using the new `table` tag:
+
+[source,xml]
+.index.jelly
+----
+<j:jelly xmlns:j="jelly:core" xmlns:dt="/data-tables" >
+    [...]
+    <st:adjunct includes="io.jenkins.plugins.data-tables"/>
+
+    <dt:table model="${it.getTableModel('id')}"/> <1>
+    [...]
+</j:jelly>
+----
+<1> replace `id` with the id of your table
+
+The only parameter you need to provide for the table is the model -- it is typically part of the corresponding
+Jenkins view model class (this object is referenced with `+${it}+` in the view).
+In order to connect the corresponding Jenkins view model class with the table, the view model class needs to
+implement the `AsyncTableContentProvider` interface. Or even simpler, let your view model class derive from
+`DefaultAsyncTableContentProvider`. This relationship is required, so that Jenkins can automatically create
+and bind a proxy for the Ajax calls that will automatically fill the table content after the HTML page has been created.
+
+If we put all those pieces together, we are required to define a model similar to the model of the Forensics plugin,
+that is shown in <<jenkins-view-model>>.
+
+[#jenkins-view-model]
+.Jenkins reporter design - high level view of the model for reporter plugins
+image::forensics-view-model.png[Forensics view model]
+
+As already described in <<jenkins-reporter-model>> the plugin needs to attach a `BuildAction` to each build. The
+Forensics plugin attaches a `ForensicBuildAction` to the build. This action stores a `RepositoryStatistics` instance,
+that contains the repository results for a given build. This action delegates all Stapler requests to a new
+https://stapler.kohsuke.org/apidocs/org/kohsuke/stapler/StaplerProxy.html[Stapler proxy instance] so we can keep the
+action clean of user interface code. This `ForensicsViewModel` class then acts as view model that provides the server
+side model for the corresponding Jelly view given by the file `index.jelly`.
+
+While this approach looks quite complex at a first view, you will see that the actual implementation part
+is quite small. Most of the boilerplate code is already provided by the base classes and you need to implement
+only a few methods. Using this concept also provides some additional features, that are part of the DataTables plugin:
+
+- Ordering of columns is persisted automatically in the browser local storage.
+- Paging size is persisted automatically in the browser local storage.
+- The Ajax calls are actually invoked only if a table will become visible. So if you have
+several tables hidden in tabs then the content will be loaded on demand only, reducing the amount of data
+to be transferred.
+- There is an option available to provide an additional details row that can be expanded with a + symbol,
+see https://github.com/jenkinsci/warnings-ng-plugin/blob/master/doc/images/details.png[warnings plugin table] for details.
```

Comment:
The image that is linked here provides a really great example of how the interface will look.  Could a link also be provided that provides more guidance on how to create that interface?

Not a blocking comment, just a question for your consideration.

Reply:
Do you mean the tabs that I am using here? Or just the overall experience using the tables and cards?